### PR TITLE
fix: make missing daemon errors actionable

### PR DESCRIPTION
**Pull request by Kylejeong2**
## Summary

- translate daemon socket `ECONNREFUSED` and `ENOENT` failures into a human-readable error
- print the exact `browse open` command that restarts the requested session
- document that API keys are forwarded to an already-running daemon and that `browse stop` is idempotent
- preserve regression coverage for late environment variables and exit-code-zero cleanup

## Root cause

The daemon client passed raw Unix socket errors through to users when the daemon disappeared between the readiness check and the request. Agents received `ECONNREFUSED` without a recovery command.

## Impact

Agents now get an actionable `daemon_not_running` failure with the exact command needed to restart the session. Recovery command arguments are shell-quoted, and a daemon disappearing between status and stop is treated as an already-stopped session. Cleanup synchronizes with daemon startup so it preserves a replacement daemon that starts during the stop race. The bundled SKILL.md also makes the already-fixed env timing and stop behavior explicit.

## E2E Test Matrix

| Command / flow | Observed output | Confidence / sufficiency |
| --- | --- | --- |
| `pnpm --filter browse build` | TypeScript compilation and oclif manifest generation completed successfully. | Proves the exact local CLI code under review builds; does not exercise a live browser. |
| `pnpm --filter browse test:cli` | 25 test files passed; 366/366 tests passed. This includes deterministic daemon disappearance/restart races, adversarial recovery-command shell quoting, malformed CDP log payloads, and the complete CLI contract suite. | Proves the full CLI suite passes against the completed local build on macOS; GitHub CI provides the Ubuntu/Windows matrix. |
| `BROWSE_DAEMON_DIR=<temp dir> node packages/cli/bin/run.js stop --session no-daemon-smoke` | Exited `0` and printed `{ "stopped": false, "session": "no-daemon-smoke" }`. | Exercises the built CLI's real absent-daemon stop path and confirms its documented successful no-op behavior. |
| `pnpm --filter browse lint` | Full-package Prettier, ESLint, and TypeScript checks completed successfully. | Reproduces the CI lint pipeline locally and verifies the daemon changes, log reducer, tests, and inherited CLI lint baseline. |

Linear: [GRO-1908](https://linear.app/browserbase/issue/GRO-1908/cli-namespace-2027-dev-suggested-docs-updates)

---
## Summary by cubic
Replaces raw socket errors with a clear `daemon_not_running` message and prints the exact, shell-quoted `browse open` recovery command. `browse stop` now treats a missing daemon as already stopped, validates ownership before cleanup, and avoids tearing down a daemon that restarts mid-race. Addresses Linear: https://linear.app/browserbase/issue/GRO-1908/cli-namespace-2027-dev-suggested-docs-updates.

- **Bug Fixes**
  - Map `ECONNREFUSED`/`ENOENT` to `daemon_not_running` and show the precise, session-aware recovery command.
  - Make `browse stop` idempotent and safe: if the daemon vanished (no `--force`), return `{ stopped: false }`; acquire a lock and verify daemon ownership before removing pid/socket/lock; preserve a replacement daemon started during the stop race (handles PID reuse).
  - Harden CDP log reduction: stricter parsing, ignore malformed payloads, and require numeric response statuses.
  - Docs: clarify `BROWSERBASE_API_KEY` is forwarded on every command and `browse stop` is idempotent; add troubleshooting for the new error.
  - Tests: cover the error message and shell-quoted recovery command (round-tripped through `/bin/sh`), stop race cleanup and restart preservation, PID reuse, and reduce-logs edge cases.

<sup>Written for commit 030c02ae69c9648e4fa3a92ff3af77a0133e50b6. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/browserbase/stagehand/pull/2356?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: packages/cli/src/lib/driver/daemon/client.ts [hunk 1]
```diff
@@ -362,3 +366,31 @@ async function removeStaleLock(lockPath: string): Promise<boolean> {
...
+    `Driver daemon session "${session}" is not running. Start it with: ${startCommand}`,
+    1,
+    { resultCode: "daemon_not_running" },
+  );
+}
+
+function formatCommandArgument(value: string): string {
+  return /^[A-Za-z0-9_./:?&=%#@+~-]+$/.test(value)
+    ? value
+    : JSON.stringify(value);
```
[the remark this answers is not in the document]
Comment (shrey150) [reply packages/cli/src/lib/driver/daemon/client.ts]: Addressed in [54dcb22bf](https://github.com/browserbase/stagehand/commit/54dcb22bf19880d02a0218b065ead63a27a56d13) -- use POSIX shell-safe quoting for URL and session arguments, with shell round-trip coverage for substitutions, whitespace, and quote characters.

File: packages/cli/src/lib/driver/daemon/client.ts [hunk 2]
```diff
@@ -225,6 +225,10 @@ async function sendDriverRequest<T>(
       }
     });
     socket.on("error", (error) => {
+      if (isDaemonUnavailableError(error)) {
+        failRequest(daemonNotRunningError(session, request));
```
[the remark this answers is not in the document]
Comment (shrey150) [reply packages/cli/src/lib/driver/daemon/client.ts:237]: Addressed in [54dcb22bf](https://github.com/browserbase/stagehand/commit/54dcb22bf19880d02a0218b065ead63a27a56d13) -- preserve idempotent stop behavior when the daemon disappears between the status and stop requests, clean stale daemon files, and cover the race with a regression test.

File: packages/cli/src/lib/driver/daemon/client.ts [hunk 3]
```diff
@@ -120,6 +120,14 @@ export async function stopDriverDaemon(
       type: "stop",
     });
   } catch (error) {
+    if (
+      !force &&
+      error instanceof CommandFailure &&
+      error.telemetry.resultCode === "daemon_not_running"
+    ) {
+      await cleanupDaemonFiles(session);
```
[the remark this answers is not in the document]
Comment (shrey150) [reply packages/cli/src/lib/driver/daemon/client.ts]: Addressed in [7312a6f31](https://github.com/browserbase/stagehand/commit/7312a6f31) -- serialize non-force stale cleanup with the per-session startup lock, recheck daemon liveness after acquiring it, and preserve a concurrently restarted daemon's fresh PID/socket. Added a deterministic restart-race regression test.

File: packages/cli/src/lib/cloud/reduce-logs.ts [hunk 4]
```diff
@@ -25,13 +40,24 @@ function paramsOf(e: RawLog): Record<string, any> {
...
+      l
+        .replace(/https?:\/\/[^/)]+/g, "")
+        .replace(/\?[^):]*/, "")
+        .trim(),
+    )
     .join("\n");
 }
 
-export function reduceLogs(raw: RawLog[], opts: ReduceLogsOptions = {}): unknown[] {
+export function reduceLogs(
```
[the remark this answers is not in the document]
Comment (shrey150) [reply packages/cli/src/lib/cloud/reduce-logs.ts:57]: Addressed in [7312a6f31](https://github.com/browserbase/stagehand/commit/7312a6f31) -- add focused reducer coverage for console, log-entry, response-status, exception, loading-failure, and malformed payload paths, plus runtime guards for malformed console argument containers and entries.

File: packages/cli/src/lib/driver/daemon/client.ts [hunk 5]
```diff
@@ -120,6 +120,14 @@ export async function stopDriverDaemon(
       type: "stop",
     });
   } catch (error) {
+    if (
+      !force &&
+      error instanceof CommandFailure &&
+      error.telemetry.resultCode === "daemon_not_running"
+    ) {
+      await cleanupStoppedDaemonFiles(session);
```
[the remark this answers is not in the document]
Comment (shrey150) [reply packages/cli/src/lib/driver/daemon/client.ts:128]: Addressed in [030c02ae6](https://github.com/browserbase/stagehand/commit/030c02ae69c9648e4fa3a92ff3af77a0133e50b6) -- replace PID-only retention with a session status handshake under the startup lock. A live unrelated process that reused the stale PID no longer preserves dead daemon files; a concurrently restarted daemon must answer the private status protocol before its files are retained. Added deterministic PID-reuse and restart-validation coverage.

**Review by shrey150 — APPROVED** [review]
